### PR TITLE
GODRIVER-2328 Upgrade all Evergreen CI test runners to Go 1.17

### PR DESCRIPTION
**Pull request by matthewdale**
The latest version of `golangci-lint` has dropped support for Go 1.16. Update the Evergreen CI runner to use Go 1.17.

Changes:
* Update the version of Go used in the Evergreen CI runner from 1.16 to 1.17
* Update contributing documentation to say Go 1.17 is required for development.
* Fix intermittently failing TLS timeout connection test that fails much more frequently on Go 1.17.
* Update Go 1.10 build to only build the "production" driver, not the test code.
* Remove now unused `hangingTLSConn` test type.

### CODE REVIEW

File: .evergreen/config.yml [hunk 1]
```diff
@@ -1919,20 +1919,20 @@ axes:
           - windows-64-vs2017-test
         variables:
           GCC_PATH: "/cygdrive/c/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin"
-          GO_DIST: "C:\\golang\\go1.16"
+          GO_DIST: "C:\\golang\\go1.17"
           PYTHON3_BINARY: "C:/python/Python38/python.exe"
           VENV_BIN_DIR: "Scripts"
       - id: "ubuntu1604-64-go-1-16"
```
Comment (benjirewis) [inline .evergreen/config.yml:1925]: For every value that you're changing to use a `GO_DIST` of `go1.17`, I think you'll want to change the `id` field to end with `go-1-17`.

File: .evergreen/config.yml [hunk 2]
```diff
@@ -1900,7 +1900,7 @@ axes:
         display_name: "Ubuntu 14.04"
         run_on: ubuntu1404-test
         variables:
-          GO_DIST: "/opt/golang/go1.16"
+          GO_DIST: "/opt/golang/go1.17"
           # The GCC toolchain that comes with Ubuntu 14.04 is incompatible with the standard binary
           # release of Go 1.16. As a result, Cgo builds on Ubuntu 14.04 fail with a linking error
```
Comment (benjirewis) [inline .evergreen/config.yml:1905]: Is this still true of Go 1.17? (I bet it is, but we'll just want to update this comment).

File: x/mongo/driver/topology/connection_test.go [hunk 3]
```diff
@@ -251,28 +251,22 @@ func TestConnection(t *testing.T) {
 						// Ensure the TLS handshake can be timed out and the connection propagates the error from the
 						// tlsConn in this case.
 
-						var hangingTLSConnectionSource tlsConnectionSourceFn = func(nc net.Conn, cfg *tls.Config) tlsConn {
-							tlsConn := tls.Client(nc, cfg)
-							return newHangingTLSConn(tlsConn, tc.maxConnectTime)
-						}
+						// Start a TCP listener on a random port and use the listener address as the
```
Comment (benjirewis) [inline x/mongo/driver/topology/connection_test.go:254]: Awesome, nice refactor!

File: .evergreen/config.yml (hunk 1 above)
Comment (matthewdale) [reply .evergreen/config.yml:1925]: Good catch, will update.

File: .evergreen/config.yml (hunk 2 above)
Comment (matthewdale) [reply .evergreen/config.yml:1905]: I would assume so, but I can test by removing the `GO_BUILD_TAGS` override to see if it fails. If it still fails, I'll update the comment to say "Go 1.16+".

**Review by prestonvasquez — APPROVED** [review]
LGTM!

File: .evergreen/config.yml (hunk 2 above)
Comment (benjirewis) [reply .evergreen/config.yml:1905]: The Ubuntu 14.04 tasks do seem to pass with that line commented out; so you should be able to just remove this comment and the `GO_BUILD_TAGS` 🤔 .

**Review by benjirewis — APPROVED** [review]
Thanks!

File: .evergreen/config.yml (hunk 2 above)
Comment (matthewdale) [reply .evergreen/config.yml:1906]: Yeah, I thought that was a bit odd so I looked into it further and realized Go 1.17 isn't available on the Ubuntu 14.04 image, so Go test was running with Go 1.7. It's useful to note that there are other Go installations in the `PATH`, so it's easy to unintentionally get the wrong Go version if something is messed up or missing.

As far as a fix, I've just reverted the Ubuntu 14.04 build to use Go 1.16, added a comment about why, and left the overridden `GO_BUILD_TAGS`.

Comment (benjirewis) [reply .evergreen/config.yml:1906]: Sounds good!

**Pull request merged**